File: Python/108. Convert Sorted Array to Binary Search Tree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        return self.build(nums, 0, len(nums))

    def build(self, nums, left, right):
        if left >= right:
            return None
        mid = left + (right - left) // 2
        node = TreeNode(nums[mid])
        node.left = self.build(nums, left, mid)
        node.right = self.build(nums, mid + 1, right)
        return node


result = Solution().sortedArrayToBST([-10,-3,0,5,9])
print(result)

File: Swift/Algorithm/Algorithm/LeetCode/108. Convert Sorted Array to Binary Search Tree.swift
//
//  108. Convert Sorted Array to Binary Search Tree.swift
//  Algorithm
//
//  Created by amoyio on 2018/10/6.
//  Copyright © 2018 amoyio. All rights reserved.
//

import Foundation
class P108 {
    class Solution {
        func sortedArrayToBST(_ nums: [Int]) -> TreeNode? {
            return build(nums, left: 0, right: nums.count)
        }
        
        func build(_ nums:[Int], left:Int, right: Int) -> TreeNode? {
            guard left < right else { return nil }
            let mid = left + (right - left) / 2
            let node = TreeNode(nums[mid])
            node.left = build(nums, left: left, right: mid)
            node.right = build(nums, left: mid + 1, right: right)
            return node
        }
    }
}
